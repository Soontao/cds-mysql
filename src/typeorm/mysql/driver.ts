/* eslint-disable max-len */
import { defaultTo } from "@newdash/newdash/defaultTo";
import { isEmpty } from "@newdash/newdash/isEmpty";
import type { ReplicationMode, TableColumn } from "typeorm";
import { MysqlDriver } from "typeorm/driver/mysql/MysqlDriver";
import { ColumnMetadata } from "typeorm/metadata/ColumnMetadata";
import { DateUtils } from "typeorm/util/DateUtils";
import { OrmUtils } from "typeorm/util/OrmUtils";
import { CDSMySQLQueryRunner } from "./query-runner";
import { CDSMySQLSchemaBuilder } from "./schema-builder";

/**
 * @internal
 */
export class CDSMySQLDriver extends MysqlDriver {
  /**
   * Creates a schema builder used to build and sync a schema.
   */
  createSchemaBuilder() {
    return new CDSMySQLSchemaBuilder(this.connection);
  }

  /**
   * Creates a query runner used to execute database queries.
   */
  createQueryRunner(mode: ReplicationMode) {
    return new CDSMySQLQueryRunner(this, mode);
  }

  normalizeType(column: ColumnMetadata) {
    if (column instanceof String) {
      return "nvarchar";
    }
    if (column.type === "uuid") {
      return "nvarchar";
    }
    if (column.type === "nvarchar" ||
      column.type === "national varchar") {
      return "nvarchar";
    }
    return super.normalizeType(column);
  }

  /**
   * Differentiate columns of this table and columns from the given column metadatas columns
   * and returns only changed.
   */
  findChangedColumns(tableColumns: TableColumn[], columnMetadataList: ColumnMetadata[]): ColumnMetadata[] {
    return columnMetadataList.filter((columnMetadata) => {
      const tableColumn = tableColumns.find((c) => c.name === columnMetadata.databaseName);
      if (!tableColumn) return false; // we don't need new columns, we only need exist and changed

      let columnMetadataLength = columnMetadata.length;
      if (!columnMetadataLength && columnMetadata.generationStrategy === "uuid") {
        // fixing #3374
        columnMetadataLength = this.getColumnLength(columnMetadata);
      }

      if (tableColumn.name !== columnMetadata.databaseName) {
        return true;
      }
      if (tableColumn.type !== this.normalizeType(columnMetadata)) {
        return true;
      }
      if (this.withLengthColumnTypes.includes(tableColumn.type as any)) {
        // for the empty length, it means its maybe a default length
        if (
          !isEmpty(tableColumn.length)
          && !isEmpty(columnMetadataLength)
          && tableColumn.length !== columnMetadataLength) {
          return true;
        }
      }

      if (this.withWidthColumnTypes.includes(tableColumn.type as any)) {
        if (tableColumn.width !== columnMetadata.width) {
          return true;
        }
      }

      if (this.withPrecisionColumnTypes.includes(tableColumn.type as any)) {
        if (columnMetadata.precision !== undefined && tableColumn.precision !== columnMetadata.precision) {
          return true;
        }
      }

      if (this.withScaleColumnTypes.includes(tableColumn.type as any)) {
        if (columnMetadata.scale !== undefined && tableColumn.scale !== columnMetadata.scale) {
          return true;
        }
      }

      if (tableColumn.zerofill !== columnMetadata.zerofill) {
        return true;
      }
      if (tableColumn.unsigned !== columnMetadata.unsigned) {
        return true;
      }
      if (tableColumn.asExpression !== columnMetadata.asExpression) {
        return true;
      }
      if (tableColumn.generatedType !== columnMetadata.generatedType) {
        return true;
      }
      if (defaultTo(tableColumn.comment, "") !== defaultTo(columnMetadata.comment, "")) {
        return true;
      }
      if (
        !this.compareDefaultValues(
          this.normalizeDefault(columnMetadata),
          tableColumn.default === "'NULL'" ? undefined : tableColumn.default
        )
      ) {
        return true;
      }
      if (
        tableColumn.enum &&
        columnMetadata.enum &&
        !OrmUtils.isArraysEqual(
          tableColumn.enum,
          columnMetadata.enum.map((val) => val + "")
        )
      ) {
        return true;
      }
      if (tableColumn.onUpdate !== columnMetadata.onUpdate) {
        return true;
      }
      if (tableColumn.isPrimary !== columnMetadata.isPrimary) {
        return true;
      }
      if (tableColumn.isNullable !== columnMetadata.isNullable) {
        return true;
      }
      if (tableColumn.isUnique !== this.normalizeIsUnique(columnMetadata)) {
        return true;
      }
      if (columnMetadata.generationStrategy !== "uuid" && tableColumn.isGenerated !== columnMetadata.isGenerated) {
        return true;
      }

      return false;
    });
  }

  /**
   * Normalizes "default" value of the column.
   */
  normalizeDefault(columnMetadata: ColumnMetadata): string | undefined {
    const defaultValue = columnMetadata.default;

    if (defaultValue === null) {
      return undefined;
    } else if (defaultValue === "'NULL'") {
      return undefined;
    } else if (
      (columnMetadata.type === "enum" || columnMetadata.type === "simple-enum" || typeof defaultValue === "string") &&
      defaultValue !== undefined
    ) {
      return `'${defaultValue}'`;
    } else if (columnMetadata.type === "set" && defaultValue !== undefined) {
      return `'${DateUtils.simpleArrayToString(defaultValue)}'`;
    } else if (typeof defaultValue === "number") {
      return `'${defaultValue.toFixed(columnMetadata.scale)}'`;
    } else if (typeof defaultValue === "boolean") {
      return defaultValue === true ? "1" : "0";
    } else if (typeof defaultValue === "function") {
      return defaultValue();
    } else {
      return defaultValue as any;
    }
  }
}
