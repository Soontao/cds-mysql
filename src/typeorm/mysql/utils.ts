
/**
 * equal string without case sensitive
 * 
 * @param s1 first string
 * @param s2 second string
 * @returns true if both two string are equal without case sensitive
 */
export function equalWithoutCase(s1: string, s2: string) {
  if (s1 === undefined && s2 === undefined) {
    return true;
  }
  if (s1?.length === 0 && s2?.length === 0) {
    return true;
  }
  if(s1 === s2) {
    return true;
  }
  if (s1?.length === s2?.length && s1?.toLowerCase() === s2?.toLowerCase()) {
    return true;
  }

  return false;
}