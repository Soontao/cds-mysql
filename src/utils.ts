
/**
 * overwrite cds compiler type mapping
 */
export function overwriteCDSCoreTypes() {

  const { cdsToSqlTypes } = require("@sap/cds-compiler/lib/render/utils/common");

  cdsToSqlTypes.sqlite = {
    "cds.Binary": "CHAR",
    "cds.hana.BINARY": "CHAR",
    "cds.hana.SMALLDECIMAL": "DECIMAL",
  };

}