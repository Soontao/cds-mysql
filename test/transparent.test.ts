/* eslint-disable max-len */
import { setupTest } from "cds-internal-tool";
import path from "node:path";
import { Query } from "typeorm/driver/Query";
import { createSh, doAfterAll, loadCSN, loadMigrateStepCSN } from "./utils";
import fs from "node:fs/promises";


describe("transparent Test Suite", () => {
  const client = setupTest(__dirname, "./resources/transparent");
  client.defaults.auth = { username: "alice", password: "admin" };

  afterAll(doAfterAll);

  it("should support get metadata", async () => {
    const response = await client.get("/fiori/$metadata");
    expect(response.status).toBe(200);
    expect(response.data).toMatch(/Persons/);
  });

  it("should support generate transparent migration script", async () => {
    const { build_migration_scripts_from_csn } = require("../src/typeorm/transparent");

    const files = await fs.readdir(path.join(__dirname, "./resources/migrate"));

    // do migration one by one
    for (let idx = 1; idx <= files.length; idx++) {
      const migration_id = `${idx - 1}->${idx}`;
      const current_csn = await loadCSN(`./resources/migrate/step-${idx}.cds`);
      const previous_csn = idx > 1 ? await loadCSN(`./resources/migrate/step-${idx - 1}.cds`) : undefined;
      const queries: Array<Query> = await build_migration_scripts_from_csn(current_csn, previous_csn);
      expect(queries.map(q => q.query).join("\n")).toMatchSnapshot(`transparent migration - ${migration_id}`);
    }

  });

  it("should never drop column when only resize the column", async () => {
    const { build_migration_scripts_from_csn } = require("../src/typeorm/transparent");
    const queries: Array<Query> = await build_migration_scripts_from_csn(
      await loadMigrateStepCSN(14), // new
      await loadMigrateStepCSN(13), // old
    );
    expect(queries).toMatchSnapshot();
  });

  it("should raise error when extend extension", async () => {

    const sh = createSh({
      cwd: path.join(__dirname, "./resources/_fiori_ext_"),
      env: process.env,
      shell: true,
      stdPipe: false
    });

    await sh("npx", "cds", "pull", "-u", "theo-on-tenant-2:pass", "--from", client.defaults.baseURL);
    await sh("npx", "cds", "build", "--for", "mtx-extension");
    const code = await sh(
      "npx", "cds", "push", "./gen/extension.tgz", "-u", "theo-on-tenant-2:pass", "--to",
      client.defaults.baseURL
    );

    expect(code).toBe(1); // should failed

    const response = await client.get("/fiori/$metadata", {
      auth: {
        username: "theo-on-tenant-2",
        password: "pass"
      }
    });

    expect(response.data).not.toMatch(/zz_ExtValue/);

  });
});