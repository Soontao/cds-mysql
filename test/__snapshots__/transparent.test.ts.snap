// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transparent Test Suite should never drop column when only resize the column 1`] = `
[
  Query {
    "@instanceof": Symbol(Query),
    "parameters": undefined,
    "query": "ALTER TABLE \`test_resources_migrate_Product\` MODIFY COLUMN \`Name\` varchar(40) NULL",
  },
]
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 0->1 1`] = `"CREATE TABLE \`test_resources_migrate_People\` (\`ID\` varchar(36) NOT NULL, \`Name\` varchar(100) NULL, PRIMARY KEY (\`ID\`)) ENGINE=InnoDB CHARACTER SET 'utf8mb4' COLLATE 'utf8mb4_unicode_ci'"`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 1->2 1`] = `
"ALTER TABLE \`test_resources_migrate_People\` ADD \`Age\` int NULL
ALTER TABLE \`test_resources_migrate_People\` DROP PRIMARY KEY
ALTER TABLE \`test_resources_migrate_People\` ADD PRIMARY KEY (\`ID\`, \`Name\`)
ALTER TABLE \`test_resources_migrate_People\` CHANGE \`Name\` \`Name\` varchar(100) NOT NULL"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 2->3 1`] = `
"ALTER TABLE \`test_resources_migrate_People\` ADD \`Active\` tinyint NULL DEFAULT 0
CREATE VIEW \`test_resources_migrate_ActivePeople\` AS SELECT
  People_0.ID,
  People_0.Name,
  People_0.Age,
  People_0.Active
FROM test_resources_migrate_People AS People_0
WHERE People_0.Active = TRUE;"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 3->4 1`] = `
"DROP VIEW \`test_resources_migrate_ActivePeople\`
CREATE VIEW \`test_resources_migrate_InActivePeople\` AS SELECT
  People_0.Name
FROM test_resources_migrate_People AS People_0
WHERE People_0.Active = FALSE;
CREATE VIEW \`test_resources_migrate_ActivePeople\` AS SELECT
  People_0.Name
FROM test_resources_migrate_People AS People_0
WHERE People_0.Active = TRUE;
CREATE VIEW \`test_resources_migrate_AllPeoples\` AS SELECT
  ActivePeople_0.Name
FROM test_resources_migrate_ActivePeople AS ActivePeople_0
UNION SELECT
  InActivePeople_1.Name
FROM test_resources_migrate_InActivePeople AS InActivePeople_1;"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 4->5 1`] = `
"CREATE TABLE \`test_resources_migrate_Job\` (\`ID\` varchar(36) NOT NULL, \`Title\` varchar(255) NULL, \`Level\` int NULL, \`Active\` tinyint NULL DEFAULT 0, PRIMARY KEY (\`ID\`)) ENGINE=InnoDB CHARACTER SET 'utf8mb4' COLLATE 'utf8mb4_unicode_ci'
CREATE VIEW \`test_resources_migrate_ActiveJobs\` AS SELECT
  Job_0.ID,
  Job_0.Title,
  Job_0.Level,
  Job_0.Active
FROM test_resources_migrate_Job AS Job_0
WHERE Job_0.Active = TRUE;"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 5->6 1`] = `
"DROP VIEW \`test_resources_migrate_AllPeoples\`
CREATE VIEW \`test_resources_migrate_AllPeoples\` AS SELECT
  People_0.Name
FROM test_resources_migrate_People AS People_0;"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 6->7 1`] = `
"DROP VIEW \`test_resources_migrate_ActiveJobs\`
ALTER TABLE \`test_resources_migrate_Job\` ADD \`People_ID\` varchar(36) NULL
ALTER TABLE \`test_resources_migrate_Job\` ADD \`People_Name\` varchar(100) NULL
CREATE VIEW \`test_resources_migrate_ActiveJobs\` AS SELECT
  Job_0.ID,
  Job_0.Title,
  Job_0.Level,
  Job_0.Active,
  Job_0.People_ID,
  Job_0.People_Name
FROM test_resources_migrate_Job AS Job_0
WHERE Job_0.Active = TRUE;
CREATE VIEW \`test_resources_migrate_PeopleWithJob\` AS SELECT
  People_0.Name,
  People_0.Active,
  Job_1.Title,
  Job_1.Level,
  Job_1.Active AS JobActive
FROM (test_resources_migrate_People AS People_0 LEFT JOIN test_resources_migrate_Job AS Job_1 ON ((Job_1.People_ID = People_0.ID) AND (Job_1.People_Name = People_0.Name)));"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 7->8 1`] = `
"CREATE TABLE \`test_resources_migrate_Job_texts\` (\`locale\` varchar(14) NOT NULL, \`ID\` varchar(36) NOT NULL, \`Title\` varchar(255) NULL, PRIMARY KEY (\`locale\`, \`ID\`)) ENGINE=InnoDB CHARACTER SET 'utf8mb4' COLLATE 'utf8mb4_unicode_ci'
CREATE VIEW \`localized_fr_test_resources_migrate_Job\` AS SELECT
  L_0.ID,
  coalesce(localized_fr_1.Title, L_0.Title) AS Title,
  L_0.Level,
  L_0.Active,
  L_0.People_ID,
  L_0.People_Name
FROM (test_resources_migrate_Job AS L_0 LEFT JOIN test_resources_migrate_Job_texts AS localized_fr_1 ON localized_fr_1.ID = L_0.ID AND localized_fr_1.locale = 'fr');
CREATE VIEW \`localized_fr_test_resources_migrate_ActiveJobs\` AS SELECT
  Job_0.ID,
  Job_0.Title,
  Job_0.Level,
  Job_0.Active,
  Job_0.People_ID,
  Job_0.People_Name
FROM localized_fr_test_resources_migrate_Job AS Job_0
WHERE Job_0.Active = TRUE;
CREATE VIEW \`localized_de_test_resources_migrate_Job\` AS SELECT
  L_0.ID,
  coalesce(localized_de_1.Title, L_0.Title) AS Title,
  L_0.Level,
  L_0.Active,
  L_0.People_ID,
  L_0.People_Name
FROM (test_resources_migrate_Job AS L_0 LEFT JOIN test_resources_migrate_Job_texts AS localized_de_1 ON localized_de_1.ID = L_0.ID AND localized_de_1.locale = 'de');
CREATE VIEW \`localized_de_test_resources_migrate_ActiveJobs\` AS SELECT
  Job_0.ID,
  Job_0.Title,
  Job_0.Level,
  Job_0.Active,
  Job_0.People_ID,
  Job_0.People_Name
FROM localized_de_test_resources_migrate_Job AS Job_0
WHERE Job_0.Active = TRUE;
CREATE VIEW \`localized_test_resources_migrate_Job\` AS SELECT
  L_0.ID,
  coalesce(localized_1.Title, L_0.Title) AS Title,
  L_0.Level,
  L_0.Active,
  L_0.People_ID,
  L_0.People_Name
FROM (test_resources_migrate_Job AS L_0 LEFT JOIN test_resources_migrate_Job_texts AS localized_1 ON localized_1.ID = L_0.ID AND localized_1.locale = 'en');
CREATE VIEW \`localized_test_resources_migrate_ActiveJobs\` AS SELECT
  Job_0.ID,
  Job_0.Title,
  Job_0.Level,
  Job_0.Active,
  Job_0.People_ID,
  Job_0.People_Name
FROM localized_test_resources_migrate_Job AS Job_0
WHERE Job_0.Active = TRUE;
CREATE VIEW \`localized_fr_test_resources_migrate_People\` AS SELECT
  L.ID,
  L.Name,
  L.Active
FROM test_resources_migrate_People AS L;
CREATE VIEW \`localized_fr_test_resources_migrate_PeopleWithJob\` AS SELECT
  People_0.Name,
  People_0.Active,
  Job_1.Title,
  Job_1.Level,
  Job_1.Active AS JobActive
FROM (localized_fr_test_resources_migrate_People AS People_0 LEFT JOIN localized_fr_test_resources_migrate_Job AS Job_1 ON ((Job_1.People_ID = People_0.ID) AND (Job_1.People_Name = People_0.Name)));
CREATE VIEW \`localized_de_test_resources_migrate_People\` AS SELECT
  L.ID,
  L.Name,
  L.Active
FROM test_resources_migrate_People AS L;
CREATE VIEW \`localized_de_test_resources_migrate_PeopleWithJob\` AS SELECT
  People_0.Name,
  People_0.Active,
  Job_1.Title,
  Job_1.Level,
  Job_1.Active AS JobActive
FROM (localized_de_test_resources_migrate_People AS People_0 LEFT JOIN localized_de_test_resources_migrate_Job AS Job_1 ON ((Job_1.People_ID = People_0.ID) AND (Job_1.People_Name = People_0.Name)));
CREATE VIEW \`localized_test_resources_migrate_People\` AS SELECT
  L.ID,
  L.Name,
  L.Active
FROM test_resources_migrate_People AS L;
CREATE VIEW \`localized_test_resources_migrate_PeopleWithJob\` AS SELECT
  People_0.Name,
  People_0.Active,
  Job_1.Title,
  Job_1.Level,
  Job_1.Active AS JobActive
FROM (localized_test_resources_migrate_People AS People_0 LEFT JOIN localized_test_resources_migrate_Job AS Job_1 ON ((Job_1.People_ID = People_0.ID) AND (Job_1.People_Name = People_0.Name)));"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 8->9 1`] = `""`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 9->10 1`] = `"CREATE TABLE \`test_resources_migrate_Product\` (\`ID\` varchar(36) NOT NULL, \`Name\` varchar(255) NULL, \`Price\` decimal(13,2) NULL, INDEX \`ProductName\` (\`Name\`), PRIMARY KEY (\`ID\`)) ENGINE=InnoDB CHARACTER SET 'utf8mb4' COLLATE 'utf8mb4_unicode_ci'"`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 10->11 1`] = `""`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 11->12 1`] = `"ALTER TABLE \`test_resources_migrate_Product\` ADD \`Description\` varchar(500) NULL"`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 12->13 1`] = `"CREATE INDEX \`ProductDescription\` ON \`test_resources_migrate_Product\` (\`Description\`)"`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 13->14 1`] = `"ALTER TABLE \`test_resources_migrate_Product\` MODIFY COLUMN \`Name\` varchar(40) NULL"`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 14->15 1`] = `
"ALTER TABLE \`test_resources_migrate_Product\` MODIFY COLUMN \`Name\` varchar(50) NULL
ALTER TABLE \`test_resources_migrate_Product\` CHANGE \`Price\` \`Price\` decimal(14,3) NULL"
`;

exports[`transparent Test Suite should support generate transparent migration script: transparent migration - 15->16 1`] = `
"ALTER TABLE \`test_resources_migrate_Product\` CHANGE \`Name\` \`Name\` varchar(50) NULL DEFAULT 'unknown product'
ALTER TABLE \`test_resources_migrate_Product\` CHANGE \`Price\` \`Price\` decimal(14,3) NULL DEFAULT '10.150'"
`;
